File: altgraph_data_app/src/main/java/org/cjoakim/cosmos/altgraph/data/processor/ImdbRawDataWranglerProcess.java
package org.cjoakim.cosmos.altgraph.data.processor;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.extern.slf4j.Slf4j;
import org.cjoakim.cosmos.altgraph.data.common.model.imdb.IndexDocument;
import org.cjoakim.cosmos.altgraph.data.common.model.imdb.Movie;
import org.cjoakim.cosmos.altgraph.data.common.model.imdb.Person;
import org.cjoakim.cosmos.altgraph.data.common.model.imdb.Principal;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.*;

/**
 * Batch process to "wrangle" the raw IMDb data into correlated movies.json and people.json
 * files.
 * <p>
 * Chris Joakim, Microsoft, November 2022
 */

@Data
@EqualsAndHashCode(callSuper = false)
@Slf4j
public class ImdbRawDataWranglerProcess extends AbstractConsoleAppProcess {

    private static int DISPLAY_BATCH_SIZE = 100000;
    int minYear;
    int minMinutes;

    long totalMovieCount = 0;
    long totalPersonCount = 0;
    long totalPrincipalCount = 0;
    long titleToPersonCount = 0;
    long startMs = 0;
    private HashMap<String, Movie> movies = new HashMap<String, Movie>();  // key is tconst
    private HashSet<String> principalSet = new HashSet<String>();
    private HashMap<String, Person> people = new HashMap<String, Person>();  // key is nconst

    public void process() throws Exception {

        Movie.setMinMinutes(minMinutes);

        startMs = System.currentTimeMillis();
        checkMemory(true, true, "start");

        readFilterMovies(minYear);
        checkMemory(true, true, "after readFilterMovies");

        readIdentifyPrincipalsInMovies();
        checkMemory(true, true, "after readIdentifyPrincipalsInMovies");

        readNamesOfPrincipals();
        checkMemory(true, true, "after readNamesOfPrincipals");

        associateTitlesToPeople();
        checkMemory(true, true, "after associateTitlesToPeople");

        writeMovies();
        checkMemory(true, true, "after writeMovies");

        writePeople();
        checkMemory(true, true, "after writePeople");

        createWriteMovieIndex();
        checkMemory(true, true, "after createWriteMovieIndex");

        checkMemory(true, true, "finish");
        displayEojTotals();
    }

    private void readFilterMovies(int minYear) throws Exception {

        String path = IMDB_RAW_TITLE_BASICS_FILE;
        sysout("readFilterMovies; path: " + path);

        InputStream is = new FileInputStream(path);
        long lineNumber = 0;

        Scanner sc = new Scanner(is, StandardCharsets.UTF_8.name());
        try {
            while (sc.hasNextLine()) {
                lineNumber++;
                String line = sc.nextLine();
                String[] lineTokens = line.split("\\t");
                if (lineNumber < 2) {
                    printHeaderRow(path, line, lineTokens);
                } else {
                    Movie movie = new Movie(lineTokens);
                    totalMovieCount++;
                    if (movie.include(minYear)) {
                        movie.scrubValues();
                        movie.setCosmosDbCoordinateAttributes(movie.getTconst(), DOCTYPE_MOVIE);
                        movies.put(movie.getTconst(), movie);
                    }
                }
            }
        } catch (Throwable t) {
            t.printStackTrace();
        } finally {
            sc.close();
        }
    }

    private void readIdentifyPrincipalsInMovies() throws Exception {

        String path = IMDB_RAW_TITLE_PRINCIPALS_FILE;
        sysout("readIdentifyPrincipalsInMovies; path: " + path);

        InputStream is = new FileInputStream(path);
        long lineNumber = 0;

        Scanner sc = new Scanner(is, StandardCharsets.UTF_8.name());
        try {
            while (sc.hasNextLine()) {
                lineNumber++;
                String line = sc.nextLine();
                String[] lineTokens = line.split("\\t");
                if (lineNumber < 2) {
                    printHeaderRow(path, line, lineTokens);
                } else {
                    Principal principal = new Principal(lineTokens);
                    principal.setDoctype(DOCTYPE_PRINCIPAL);
                    totalPrincipalCount++;
                    if (principal.isValid()) {
                        if (principal.hasNecessaryRole()) {  // actress,actor,director
                            if (movies.containsKey(principal.getTconst())) {
                                principalSet.add(principal.getNconst());
                            }
                        }
                    }
                }
            }
        } catch (Throwable t) {
            t.printStackTrace();
        } finally {
            sc.close();
        }
    }

    private void readNamesOfPrincipals() throws Exception {

        String path = IMDB_RAW_NAME_BASICS_FILE;
        sysout("readNamesOfPrincipals; path: " + path);

        InputStream is = new FileInputStream(path);
        long lineNumber = 0;

        Scanner sc = new Scanner(is, StandardCharsets.UTF_8.name());
        try {
            while (sc.hasNextLine()) {
                lineNumber++;
                String line = sc.nextLine();
                String[] lineTokens = line.split("\\t");
                if (lineNumber < 2) {
                    printHeaderRow(path, line, lineTokens);
                } else {
                    Person person = new Person(lineTokens);
                    totalPersonCount++;
                    if (person.isValid()) {
                        if (principalSet.contains(person.getNconst())) {
                            person.scrubValues();
                            person.setCosmosDbCoordinateAttributes(person.getNconst(), DOCTYPE_PERSON);
                            people.put(person.getNconst(), person);
                        }
                    }
                }
            }
        } catch (Throwable t) {
            t.printStackTrace();
        } finally {
            sc.close();
        }
    }

    private void writeMovies() throws Exception {

        String path = IMDB_MOVIES_DOCUMENTS_FILE;
        sysout("writeMovies to " + path);

        try (FileOutputStream out = new FileOutputStream(path);
             OutputStreamWriter writer = new OutputStreamWriter(out)) {

            Iterator<String> iterator = movies.keySet().iterator();
            while (iterator.hasNext()) {
                String key = iterator.next();
                Movie movie = movies.get(key);
                String json = movie.asJson(false).trim();
                if (json.startsWith("{")) {
                    if (json.endsWith("}")) {
                        writer.write(json);
                        writer.write(System.lineSeparator());
                    }
                }
            }
        }
    }

    private void writePeople() throws Exception {

        String path = IMDB_PEOPLE_DOCUMENTS_FILE;
        sysout("writePeople to " + path);

        try (FileOutputStream out = new FileOutputStream(path);
             OutputStreamWriter writer = new OutputStreamWriter(out)) {

            Iterator<String> iterator = people.keySet().iterator();
            while (iterator.hasNext()) {
                String key = iterator.next();
                Person person = people.get(key);
                String json = person.asJson(false).trim();
                if (json.startsWith("{")) {
                    if (json.endsWith("}")) {
                        writer.write(json);
                        writer.write(System.lineSeparator());
                    }
                }
            }
        }
    }

    private void associateTitlesToPeople() throws Exception {

        String path = IMDB_RAW_TITLE_PRINCIPALS_FILE;
        sysout("associateTitlesToPeople; path: " + path);

        InputStream is = new FileInputStream(path);
        long lineNumber = 0;
        Scanner sc = new Scanner(is, StandardCharsets.UTF_8.name());
        try {
            while (sc.hasNextLine()) {
                String line = sc.nextLine();
                lineNumber++;
                if (lineNumber > 1) {
                    String[] lineTokens = line.split("\\t");
                    Principal principal = new Principal(lineTokens);
                    if (principal.isValid()) {
                        String nconst = principal.getNconst();
                        if (people.containsKey(nconst)) {
                            String tconst = principal.getTconst();
                            if (movies.containsKey(tconst)) {
                                Person person = people.get(nconst);
                                Movie movie = movies.get(tconst);
                                movie.addPerson(nconst);
                                person.addTitle(tconst);
                                titleToPersonCount++;
                            }
                        }
                    }
                }
            }
        } catch (Throwable t) {
            t.printStackTrace();
        } finally {
            sc.close();
        }
    }

    private void createWriteMovieIndex() throws Exception {

        String path = IMDB_MOVIES_INDEX_FILE;
        ArrayList<IndexDocument> indexDocs = new ArrayList<IndexDocument>();
        Iterator<String> moviesIt = movies.keySet().iterator();
        while (moviesIt.hasNext()) {
            String key = moviesIt.next();
            Movie movie = movies.get(key);
            IndexDocument idxDoc = new IndexDocument(
                    DOCTYPE_MOVIE_IDX, movie.getId(), movie.getPk());

            Iterator<String> peopleIt = movie.getPeople().iterator();
            while (peopleIt.hasNext()) {
                idxDoc.addAdjacentVertex(peopleIt.next());
            }
            idxDoc.intern();
            indexDocs.add(idxDoc);
        }

        try (FileOutputStream out = new FileOutputStream(path);
             OutputStreamWriter writer = new OutputStreamWriter(out)) {
            for (int i = 0; i < indexDocs.size(); i++) {
                IndexDocument idxDoc = indexDocs.get(i);
                String json = idxDoc.asJson(false).trim();
                if (json.startsWith("{")) {
                    if (json.endsWith("}")) {
                        writer.write(json);
                        writer.write(System.lineSeparator());
                    }
                }
            }
        }
        sysout("createWriteMovieIndex, file written: " + path + "  docs: " + indexDocs.size());
    }

    private void displayEojTotals() {

        double elapsedMs = System.currentTimeMillis() - startMs;
        double elapsedSec = elapsedMs / 1000.0;
        double elapsedMin = elapsedSec / 60.0;

        sysout("");
        sysout("EOJ Totals:");
        sysout("  minYear:                  " + minYear);
        sysout("  minMinutes:               " + minMinutes);
        sysout("  totalMovieCount:          " + totalMovieCount);
        sysout("  includedMovieCount:       " + movies.size());
        sysout("  totalPrincipalCount:      " + totalPrincipalCount);
        sysout("  includedPrincipalCount:   " + principalSet.size());
        sysout("  totalPersonCount:         " + totalPersonCount);
        sysout("  includedPersonCount:      " + people.size());
        sysout("  included Movies + People: " + (people.size() + movies.size()));
        sysout("  titleToPersonCount:       " + titleToPersonCount);
        sysout("  elapsedMs:                " + elapsedMs);
        sysout("  elapsedMin:               " + elapsedMin);
    }

    private void printHeaderRow(String path, String line, String[] lineTokens) {

        sysout("path:   " + path);
        sysout("header: " + line);
        for (int t = 0; t < lineTokens.length; t++) {
            sysout("token: " + t + " -> " + lineTokens[t]);
        }
    }
}

File: src/altgraph_data_app/processor/ImdbRawDataWranglerProcess.cs
using System.Text.Json;
using altgraph_shared_app;
using altgraph_shared_app.Models.Imdb;
using altgraph_shared_app.Options;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace altgraph_data_app.processor
{
  public class ImdbRawDataWranglerProcess : IConsoleAppProcess
  {
    private readonly ImdbPathsOptions _imdbPathsOptions;
    private readonly ILogger<ImdbRawDataWranglerProcess> _logger;
    private readonly ConsoleAppProcess _consoleAppProcess;
    private long _totalMovieCount = 0;
    private long _totalPersonCount = 0;
    private long _totalPrincipalCount = 0;
    private long _titleToPersonCount = 0;
    private long _startMs = 0;
    private Dictionary<string, Movie> _movies = new Dictionary<string, Movie>();
    private Dictionary<string, Person> _people = new Dictionary<string, Person>();
    private HashSet<string> _principalSet = new HashSet<string>();

    public int MinYear { get; set; } = 0;
    public int MinMinutes { get; set; } = 0;

    public ImdbRawDataWranglerProcess(ILogger<ImdbRawDataWranglerProcess> logger, IOptions<ImdbPathsOptions> imdbPathsOptions, ConsoleAppProcess consoleAppProcess)
    {
      _logger = logger;
      _imdbPathsOptions = imdbPathsOptions.Value;
      _consoleAppProcess = consoleAppProcess;
    }

    public async Task ProcessAsync()
    {
      _startMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

      await ReadFilterMoviesAsync();

      await ReadIdentifyPrincipalsInMoviesAsync();

      await ReadNamesOfPrincipalsAsync();

      await AssociateTitlesToPeopleAsync();

      await WriteMoviesAsync();

      await WritePeopleAsync();

      await CreateWriteMovieSeedAsync();

      DisplayEojTotals();
    }

    private async Task ReadFilterMoviesAsync()
    {
      string path = _imdbPathsOptions.RawTitleBasicsFile;
      _logger.LogDebug($"ReadFilterMoviesAsync, path: {path}");

      bool isFirstLine = true;

      try
      {
        using (StreamReader sr = new StreamReader(path))
        {
          string? line = null;
          while ((line = await sr.ReadLineAsync()) != null)
          {
            if (isFirstLine)
            {
              isFirstLine = false;
              continue;
            }
            string[] lineTokens = line.Split("\t");
            Movie movie = new Movie(lineTokens);
            if (movie.Include(MinYear, MinMinutes))
            {
              movie.SetCosmosDbCoordinateAttributes(movie.TConst, Constants.DOCTYPE_MOVIE);
              _movies.Add(movie.TConst, movie);
              _totalMovieCount++;
            }
          }
        }
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, $"ReadFilterMoviesAsync, path: {path}");
      }
    }

    private async Task ReadIdentifyPrincipalsInMoviesAsync()
    {
      string path = _imdbPathsOptions.RawTitlePrincipalsFile;
      _logger.LogDebug($"ReadIdentifyPrincipalsInMoviesAsync, path: {path}");

      bool isFirstLine = true;

      try
      {
        using (StreamReader sr = new StreamReader(path))
        {
          string? line = null;
          while ((line = await sr.ReadLineAsync()) != null)
          {
            if (isFirstLine)
            {
              isFirstLine = false;
              continue;
            }

            string[] lineTokens = line.Split("\t");
            Principal principal = new Principal(lineTokens);
            principal.Doctype = Constants.DOCTYPE_PRINCIPAL;
            _totalPrincipalCount++;

            if (principal.IsValid() && principal.HasNecessaryRole() && _movies.ContainsKey(principal.TConst))
            {
              _principalSet.Add(principal.NConst);
            }
          }
        }
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, $"ReadFilterMoviesAsync, path: {path}");
      }
    }

    private async Task ReadNamesOfPrincipalsAsync()
    {
      string path = _imdbPathsOptions.RawNameBasicsFile;
      _logger.LogDebug($"ReadIdentifyPrincipalsInMoviesAsync, path: {path}");

      bool isFirstLine = true;

      try
      {
        using (StreamReader sr = new StreamReader(path))
        {
          string? line = null;
          while ((line = await sr.ReadLineAsync()) != null)
          {
            if (isFirstLine)
            {
              isFirstLine = false;
              continue;
            }

            string[] lineTokens = line.Split("\t");
            Person person = new Person(lineTokens);
            person.Doctype = Constants.DOCTYPE_PRINCIPAL;
            _totalPersonCount++;

            if (person.IsValid() && _principalSet.Contains(person.NConst))
            {
              person.ScrubValues();
              person.SetCosmosDbCoordinateAttributes(person.NConst, Constants.DOCTYPE_PERSON);
              _people.Add(person.NConst, person);
            }
          }
        }
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, $"ReadIdentifyPrincipalsInMoviesAsync, path: {path}");
      }
    }

    private async Task WriteMoviesAsync()
    {
      string path = _imdbPathsOptions.MoviesDocumentsFile;
      _logger.LogDebug($"WriteMoviesAsync, path: {path}");
      try
      {
        using (StreamWriter sw = new StreamWriter(path))
        {
          foreach (var movie in _movies)
          {
            await sw.WriteLineAsync(JsonSerializer.Serialize(movie.Value));
          }
        }
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, $"WriteMoviesAsync, path: {path}");
      }
    }

    private async Task WritePeopleAsync()
    {
      string path = _imdbPathsOptions.PeopleDocumentsFile;
      _logger.LogDebug($"WritePeopleAsync, path: {path}");
      try
      {
        using (StreamWriter sw = new StreamWriter(path))
        {
          foreach (var people in _people)
          {
            await sw.WriteLineAsync(JsonSerializer.Serialize(people.Value));
          }
        }
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, $"WritePeopleAsync, path: {path}");
      }
    }

    private async Task AssociateTitlesToPeopleAsync()
    {
      string path = _imdbPathsOptions.RawTitlePrincipalsFile;
      _logger.LogDebug($"ReadFilterMoviesAsync, path: {path}");

      bool isFirstLine = true;

      try
      {
        using (StreamReader sr = new StreamReader(path))
        {
          string? line = null;
          while ((line = await sr.ReadLineAsync()) != null)
          {
            if (isFirstLine)
            {
              isFirstLine = false;
              continue;
            }

            string[] lineTokens = line.Split("\t");
            Principal? principal = new Principal(lineTokens);
            if (principal.IsValid() && _people.ContainsKey(principal.NConst) && _movies.ContainsKey(principal.TConst))
            {
              Person person = _people[principal.NConst];
              Movie movie = _movies[principal.TConst];
              movie.AddPerson(principal.NConst);
              person.AddTitle(principal.TConst);
              _titleToPersonCount++;
            }
          }
        }
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, $"ReadFilterMoviesAsync, path: {path}");
      }
    }

    private async Task CreateWriteMovieSeedAsync()
    {
      string path = _imdbPathsOptions.MoviesSeedFile;
      _logger.LogDebug($"CreateWriteMovieSeedAsync, path: {path}");

      List<SeedDocument> indexDocs = new List<SeedDocument>();
      foreach (var movie in _movies)
      {
        Movie relatedMovie = _movies[movie.Key];
        SeedDocument seedDoc = new SeedDocument(Constants.DOCTYPE_MOVIE_SEED, relatedMovie.Id, relatedMovie.Pk);
        foreach (var person in relatedMovie.People)
        {
          seedDoc.AddAdjacentVertex(person);
        }
        indexDocs.Add(seedDoc);
      }

      try
      {
        using (StreamWriter sw = new StreamWriter(path))
        {
          foreach (var indexDoc in indexDocs)
          {
            await sw.WriteLineAsync(JsonSerializer.Serialize(indexDoc));
          }
        }
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, $"CreateWriteMovieSeedAsync, path: {path}");
      }
    }

    private void DisplayEojTotals()
    {
      double elapsedMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - _startMs;
      double elapsedSec = elapsedMs / 1000;
      double elapsedMin = elapsedSec / 60;

      _logger.LogInformation("EOJ Totals:");
      _logger.LogInformation($"Min Year: {MinYear}");
      _logger.LogInformation($"Min Minutes: {MinMinutes}");
      _logger.LogInformation($"Total Movie Count: {_totalMovieCount}");
      _logger.LogInformation($"Included Movie Count: {_movies.Count}");
      _logger.LogInformation($"Total Principal Count: {_totalPrincipalCount}");
      _logger.LogInformation($"Included Principal Count: {_principalSet.Count}");
      _logger.LogInformation($"Total Person Count: {_totalPersonCount}");
      _logger.LogInformation($"Included Person Count: {_people.Count}");
      _logger.LogInformation($"Included Movies + People: {_people.Count + _movies.Count}");
      _logger.LogInformation($"Title to Person Count: {_titleToPersonCount}");
      _logger.LogInformation($"Elapsed Time (ms): {elapsedMs}");
      _logger.LogInformation($"Elapsed Time (min): {elapsedMin}");
    }
  }
}